analogdevicesinc/pyadi-iio
# ad4020.py: add compatibility for ad400x parts

# Description

- Extend the compatible parts list to include ad4021, ad4022
- Allow passing the device name as a parameter, with ad4020 as the default
- Handle no device name or bad device name cases
- Add classes for ad4000 (supports ad4004, ad4008), ad4001 (supports ad4005), ad4002 (supports ad4006, ad4010), ad4003 (supports ad4007, ad4011)
- Import new ad400x classes in adi/__init__.py
- Add sampling_frequency setter to support variable sampling frequency
- Add test_ad4001py, test_ad4002.py, test_ad4003.py
- Add sampling_frequency tests in test_ad4020.py

## Type of change


# How has this been tested?

Tested against ad4020 (sanity check), ad4001, ad4002, ad4003 eval boards attached to a zedboard with kernel based on this PR: https://github.com/analogdevicesinc/linux/pull/2171

Note: This PR should not be merged until the above is merged into the ADI linux tree.

Generalized manual test flow:

1. install pyadi-iio and dependencies in venv
2. Start Python interpreter
3. `import adi`
4. Instantiate adc object, e.g. `adc = adi.ad4001(uri="ip:10.2.5.210")`
5. Check that `adc.rx()` succeeds with no warnings or errors
6. Check sampling frequency with `adc.sampling_frequency`
7. Set sampling frequency with `adc.sampling_frequency = <value>`
8. Check sampling frequency again with `adc.sampling_frequency` and confirm change successful

Automated tests for pytest-libiio added as test/test_ad4001.py and test_ad4002.py, and run from local machine using e.g.:
`pytest --hw=ad4001 -v test/test_ad4001.py --uri=ip:10.2.5.210`


**Test Configuration**:
* Hardware: Zedboard + EVAL-AD4001FMCZ/EVAL-AD4002FMCZ/EVAL-AD4003FMCZ/EVAL-AD4020FMCZ 
* OS: 

```root@zed-4:~# cat /etc/os-release 
PRETTY_NAME="Kuiper GNU/Linux 11.2 (bullseye)"
NAME="Kuiper GNU/Linux"
VERSION_ID="11.2"
VERSION="11.2 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="https://github.com/analogdevicesinc/adi-kuiper-gen"
SUPPORT_URL="https://ez.analog.com/linux-software-drivers/"

root@zed-4:~# uname -r
6.1.0-ad400x-dev-22246-g13a485c55367
```
Custom kernel based off of linux PR 2171 as mentioned above.

# Documentation

If this is a new feature or example please mention or link any documentation. All new hardware interface classes require documentation.

# Checklist:

- [X] I have signed off all commits and they contain "Signed-off by: <insert name>"
